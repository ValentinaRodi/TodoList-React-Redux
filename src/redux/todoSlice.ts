import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { Todo } from "../models/Todo";
import { v4 as uuidv4 } from "uuid";

const initialState = [] as Todo[];

const todoSlice = createSlice({
  name: "todos",
  initialState,
  reducers: {
    addTodo: {
      reducer: (state: any[], action: PayloadAction<Todo>) => {
        state.push(action.payload);
      },
      prepare: (description: string) => ({
        payload: {
          id: uuidv4(),
          description,
          styleBtn: "item__btn",
          styleText: "item__text",
          counter: 1,
          itemShow: 2,
        } as Todo,
      }),
    },
    removeTodo(state: any[], action: PayloadAction<string>) {
      const index = state.findIndex((todo) => todo.id === action.payload);
      state.splice(index, 1);
    },
    setTodoStatus(
      state: any[],
      action: PayloadAction<{ id: string }>
    ) {
      const index = state.findIndex((todo) => todo.id === action.payload.id);
      state[index].styleBtn = "item__btn_check";
      state[index].styleText = "item__text_check";
      state[index].counter = 0;
    },
    showItemActive(state: any[], action: PayloadAction<{ todoItemShow: number }>) {
      state.map((item)=>item.itemShow = action.payload.todoItemShow)
    },
    removeTodoAll() {
      return initialState
    },
    
  },
});

export const { addTodo, removeTodo, setTodoStatus, showItemActive, removeTodoAll } = todoSlice.actions;
export default todoSlice.reducer;


