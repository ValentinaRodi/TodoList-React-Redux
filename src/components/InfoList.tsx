import { useState, useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import { AppDispatch, RootState } from "../redux/store";
import { removeTodoAll, showItemActive } from "../redux/todoSlice";
import  './infoList.css'

export default function InfoList() {
   
    const todoList = useSelector((state: RootState) => state);
    const currentCounter = todoList.map(item => item.counter).reduce((sum, current) => sum + current, 0);
    const dispatch = useDispatch<AppDispatch>();

    //React Hooks
    const [todoItemShow, setTodoItemShow] = useState(2);
    const [todoButtonClass, setTodoButtonAllClass] = useState(['info__btn', 'info__btn', 'info__btn_check']);

    function handleChangeButtonClass() {
        const buttonTodoItem = document.querySelector<HTMLElement>('.info-item__button')! 
        buttonTodoItem.addEventListener('click', (event) => {
            const idItem = +(event.target as HTMLElement).id
            for (let index = 0; index < todoButtonClass.length; index++) {
                (index === idItem) ? todoButtonClass[index] = 'info__btn_check' : todoButtonClass[index] = 'info__btn'
            }
            setTodoButtonAllClass(todoButtonClass)     
        });
    }

    function handleClickChangeItemShow() {
        const buttonTodoItem = document.querySelector('.info-item__button')!
        buttonTodoItem.addEventListener('click', (event) => {
            const idItem = +(event.target as HTMLElement).id
            setTodoItemShow(idItem);
        }) 
    }

    function handleClickChangeItemList() {
        handleChangeButtonClass()
        handleClickChangeItemShow()
        dispatch(showItemActive({ todoItemShow: todoItemShow }));
    }
  
    //для отслеживания изменений todoItemShow
    useEffect(() => {
        handleClickChangeItemList();
    }, [todoItemShow])  

    return (
        <div>
            <div className='info-item'>
                <p className='iifo-item__counter'>{currentCounter} item left</p>
                <div className='info-item__button'>
                    <button
                        className={todoButtonClass[2]}
                        id="2"
                        onClick={handleClickChangeItemList}
                    >all
                    </button>
                    <button
                        className={todoButtonClass[1]}
                        id="1"
                        onClick={handleClickChangeItemList}
                    >active
                    </button>
                    <button
                        className={todoButtonClass[0]}
                        id="0"
                        onClick={handleClickChangeItemList}
                    >completed
                    </button>
                </div>
                <button
                    className='info__btn_clear'
                    onClick={() => {
                        dispatch(removeTodoAll())
                    }}
                >Clear completed
                </button>
            </div>
        </div>
    )
}