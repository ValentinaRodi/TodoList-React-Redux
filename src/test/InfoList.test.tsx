import  InfoList  from '../components/InfoList';
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { store } from "../redux/store";

describe("<InfoList />", () => {
  it("should render successfully and match snapshot", () => {
    const { container } = render(
        <Provider store={store}> 
            <InfoList />
        </Provider>
    );
    expect(container).toMatchSnapshot();
  });
});
