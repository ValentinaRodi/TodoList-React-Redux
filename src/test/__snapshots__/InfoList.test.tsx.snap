// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InfoList /> should render successfully and match snapshot 1`] = `
<div>
  <div>
    <div
      class="info-item"
    >
      <p
        class="iifo-item__counter"
      >
        0
         item left
      </p>
      <div
        class="info-item__button"
      >
        <button
          class="info__btn_check"
          id="2"
        >
          all
        </button>
        <button
          class="info__btn"
          id="1"
        >
          active
        </button>
        <button
          class="info__btn"
          id="0"
        >
          completed
        </button>
      </div>
      <button
        class="info__btn_clear"
      >
        Clear completed
      </button>
    </div>
  </div>
</div>
`;
